universal-application-tool-0.0.1/app/modules/SecurityModule.java

```diff
@@ -103,20 +113,79 @@ protected ProfileFactory provideProfileFactory(
     return new ProfileFactory(clock, dbContext, httpContext);
   }
 
+  @Provides
+  @Singleton
   protected FakeAdminClient fakeAdminClient(ProfileFactory profileFactory) {
     return new FakeAdminClient(profileFactory);
   }
 
+  @Named("idcs")
+  @Provides
+  @Nullable
+  @Singleton
+  protected OidcClient provideIDCSClient(ProfileFactory profileFactory) {
+    if (!this.configuration.hasPath("idcs.client_id") || !this.configuration.hasPath("idcs.secret")) {
+      return null;
+    }
+    OidcConfiguration config = new OidcConfiguration();
+    config.setClientId(this.configuration.getString("idcs.client_id"));
+    config.setSecret(this.configuration.getString("idcs.secret"));
+    config.setDiscoveryURI(this.configuration.getString("idcs.discovery_uri"));
+    config.setResponseMode("form_post");
+    config.setResponseType("id_token");
+    config.setUseNonce(true);
+    config.setWithState(false);
+    OidcClient client = new OidcClient(config);
+    client.setCallbackUrl(baseUrl + "/callback");
+    client.setProfileCreator(new IdcsProfileAdapter(config, client, profileFactory));
+    client.setCallbackUrlResolver(new PathParameterCallbackUrlResolver());
+    return client;
+  }
+
+  @Named("ad")
   @Provides
+  @Nullable
   @Singleton
-  protected Config provideConfig(GuestClient guestClient, FakeAdminClient fakeAdminClient) {
-    // This must match the line in `routes` also.
-    Clients clients = new Clients(baseUrl + routes.CallbackController.callback("GuestClient"));
+  protected OidcClient provideAdClient(ProfileFactory profileFactory) {
+    if (!this.configuration.hasPath("adfs.client_id") || !this.configuration.hasPath("adfs.secret")) {
+      return null;
+    }
+    OidcConfiguration config = new OidcConfiguration();
+    config.setClientId(this.configuration.getString("adfs.client_id"));
+    config.setSecret(this.configuration.getString("adfs.secret"));
+    config.setDiscoveryURI(this.configuration.getString("adfs.discovery_uri"));
+    config.setResponseMode("form_post");
+    config.setResponseType("id_token");
+    config.setUseNonce(true);
+    config.setWithState(false);
+    OidcClient client = new OidcClient(config);
+    client.setName("AdClient");
+    client.setCallbackUrl(baseUrl + "/callback");
+    client.setProfileCreator(new AdfsProfileAdapter(config, client, profileFactory));
+    client.setCallbackUrlResolver(new PathParameterCallbackUrlResolver());
+    return client;
+  }
+
+  @Provides
+  @Singleton
+  protected Config provideConfig(
+      GuestClient guestClient,
+      @Named("ad") @Nullable OidcClient adClient,
+      @Named("idcs") @Nullable OidcClient idcsClient,
+      FakeAdminClient fakeAdminClient) {
+    List<Client> clientList = new ArrayList<Client>();
+    clientList.add(guestClient);
+    if (idcsClient != null) {
+      clientList.add(idcsClient);
+    }
+    if (adClient != null) {
+      clientList.add(adClient);
```

Comment:
question: If these are null, does that indicate a failure to connect to the providers? Should we just fail in that case?

Reply:
It means the secrets weren't provided, which is probably the standard case in dev, so, necessary to succeed here.

Reply:
Thanks for clarifying!